crates/aiken-lang/src/parser/definition/test.rs

```diff
@@ -13,35 +19,101 @@ pub fn parser() -> impl Parser<Token, ast::UntypedDefinition, Error = ParseError
         .or_not()
         .then_ignore(just(Token::Test))
         .then(select! {Token::Name {name} => name})
-        .then_ignore(just(Token::LeftParen))
-        .then_ignore(just(Token::RightParen))
+        .then(
+            via()
+                .separated_by(just(Token::Comma))
+                .allow_trailing()
+                .delimited_by(just(Token::LeftParen), just(Token::RightParen)),
+        )
         .then(just(Token::Fail).ignored().or_not())
         .map_with_span(|name, span| (name, span))
         .then(
             expr::sequence()
                 .or_not()
                 .delimited_by(just(Token::LeftBrace), just(Token::RightBrace)),
         )
-        .map_with_span(|((((old_fail, name), fail), span_end), body), span| {
-            ast::UntypedDefinition::Test(ast::Function {
-                arguments: vec![],
-                body: body.unwrap_or_else(|| UntypedExpr::todo(None, span)),
-                doc: None,
-                location: span_end,
-                end_position: span.end - 1,
+        .map_with_span(
+            |(((((old_fail, name), arguments), fail), span_end), body), span| {
+                ast::UntypedDefinition::Test(ast::Function {
+                    arguments,
+                    body: body.unwrap_or_else(|| UntypedExpr::todo(None, span)),
+                    doc: None,
+                    location: span_end,
+                    end_position: span.end - 1,
+                    name,
+                    public: false,
+                    return_annotation: Some(ast::Annotation::boolean(span)),
+                    return_type: (),
+                    can_error: fail.is_some() || old_fail.is_some(),
+                })
+            },
+        )
+}
+
+pub fn via() -> impl Parser<Token, ast::UntypedArgVia, Error = ParseError> {
+    choice((
+        select! {Token::DiscardName {name} => name}.map_with_span(|name, span| {
+            ast::ArgName::Discarded {
+                label: name.clone(),
                 name,
-                public: false,
-                return_annotation: None,
-                return_type: (),
-                can_error: fail.is_some() || old_fail.is_some(),
+                location: span,
+            }
+        }),
+        select! {Token::Name {name} => name}.map_with_span(move |name, location| {
+            ast::ArgName::Named {
+                label: name.clone(),
+                name,
+                location,
+                is_validator_param: false,
+            }
+        }),
+    ))
+    .then(just(Token::Colon).ignore_then(annotation()).or_not())
+    .map_with_span(|(arg_name, annotation), location| (arg_name, annotation, location))
```

Comment:
Note: we do take the span _here_ to not include the `via` keyword and the following expression in it. This allows for labeling either the _via_ annotation (since it has its own location), or the argument's label depending on the error.

Reply:
clever